libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2078,6 +2086,112 @@ void TypeChecker::typeCheckABIEncodeFunctions(
 	}
 }
 
+void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCall)
+{
+	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+
+	// Expecting first argument to be the function pointer and second to be a tuple.
+	if (arguments.size() != 2)
+	{
+		m_errorReporter.typeError(
+			6219_error,
+			_functionCall.location(),
+			"Expected two arguments: a function pointer followed by a tuple."
+		);
+		return;
+	}
+
+	auto const functionPointerType = dynamic_cast<FunctionTypePointer>(type(*arguments.front()));
+
+	if (!functionPointerType)
+	{
+		m_errorReporter.typeError(
+			5511_error,
+			arguments.front()->location(),
+			"Expected first argument to be a function pointer, not \"" +
+			type(*arguments.front())->canonicalName() +
+			"\"."
+		);
+		return;
+	}
+
+	if (functionPointerType->kind() != FunctionType::Kind::External)
+	{
+		SecondarySourceLocation ssl{};
+
+		if (functionPointerType->hasDeclaration())
+			ssl.append("Function is declared here:", functionPointerType->declaration().location());
+		m_errorReporter.typeError(
+			3509_error,
+			arguments[0]->location(),
+			ssl,
+			"Function must be \"public\" or \"external\". Did you forget to precede \"this.\"?"
```

Comment:
Do you think it is a common error to miss `this.`? I.e. is it common to encode calls to the current contract?
Also I think `prefix` could be better than `precede`.

Reply:
> it common to encode calls to the current contract?

Hmm I naively used my own experience where I _did_ forget `this.` in the tests, but you're right, the common case is usually external contracts, not the current one.

I like `prefix` more, too. Or I just remove it.